15.8.1 build failure on cross-platform development teams or CI/Docker pipelines

### Current Behavior

dependency installs fails due to inability to find module `'@nrwl/nx-linux-x64-musl'` in docker containers/pipelines.

### Expected Behavior

Nx projects should install and build  with same dependencies locally as in pipeline containers so dependencies can be frozen for change control/auditing purposes.

### GitHub Repo

_No response_

### Steps to Reproduce

1. Develop on mac or windows
2. install dependencies
3. run `yarn install --ignore-optional --ignore-scripts --silent --frozen-lockfile` in a docker container


### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node : 18.12.1
   OS   : darwin arm64
   yarn : 1.22.19
   
   nx                      : 15.8.1
   @nrwl/js                : 15.8.1
   @nrwl/jest              : 15.8.1
   @nrwl/linter            : 15.8.1
   @nrwl/workspace         : 15.8.1
   @nrwl/cli               : 15.8.1
   @nrwl/cypress           : 15.8.1
   @nrwl/devkit            : 15.8.1
   @nrwl/eslint-plugin-nx  : 15.8.1
   @nrwl/node              : 15.8.1
   @nrwl/nx-plugin         : 15.8.1
   @nrwl/react             : 15.8.1
   @nrwl/rollup            : 15.8.1
   @nrwl/tao               : 15.8.1
   @nrwl/web               : 15.8.1
   @nrwl/webpack           : 15.8.1
   typescript              : 4.9.5
   ---------------------------------------
   Community plugins:
   @jscutlery/semver : 2.28.0
   ---------------------------------------
   Local workspace plugins:
         @my-project/nx

✨  Done in 1.24s.
```


### Failure Logs

```shell
Step 9/16 : RUN yarn install --ignore-optional --ignore-scripts --silent --frozen-lockfile
 ---> Running in d5b6adb840d2
Removing intermediate container d5b6adb840d2
 ---> 5b36acfb8a4f
Step 10/16 : COPY . /app
 ---> c35c6da79680
Step 11/16 : RUN yarn nx build my-project
 ---> Running in c478b6c3918e
yarn run v1.22.19
$ /app/node_modules/.bin/nx build my-project
 >  NX   Cannot find module '@nrwl/nx-linux-x64-musl'
   Require stack:
   - /app/node_modules/nx/src/native/index.js
   - /app/node_modules/nx/src/hasher/native-file-hasher.js
   - /app/node_modules/nx/src/hasher/file-hasher.js
   - /app/node_modules/nx/src/project-graph/build-project-graph.js
   - /app/node_modules/nx/src/project-graph/project-graph.js
   - /app/node_modules/nx/src/utils/project-graph-utils.js
   - /app/node_modules/nx/src/tasks-runner/utils.js
   - /app/node_modules/nx/src/tasks-runner/run-command.js
   - /app/node_modules/nx/src/command-line/run-one.js
   - /app/node_modules/nx/src/command-line/nx-commands.js
   - /app/node_modules/nx/bin/init-local.js
   - /app/node_modules/nx/bin/nx.js
error Command failed with exit code 1.
```


### Additional Information

This is probably related to the implementation of https://github.com/nrwl/nx/pull/15071, and may be a documentation issue, (no documentation was added for the 

I am going to try setting `NX_NON_NATIVE_HASHER=true` in my docker/builds to see if this fixes, but i think there was a step missing in the design considerations for this feature.

Hi, I have a similar problem `Error: Cannot find module '@nrwl/nx-linux-x64-musl'` running docker build on the CI (DigitalOcean App platform).

`Dockerfile`
```
FROM node:18.14.1-alpine as build-system
WORKDIR /app
COPY package*.json ./
RUN npm install
COPY . .
RUN nx run-many --target=build
```

- `"@nrwl/cli": "15.8.5"`
- `node:18.14.1-alpine`
- not using `--ignore-optional`
- `NX_NON_NATIVE_HASHER=true` did not helped

Thank you, @Cammisuli for trying to resolve this problem!
I still do not see the package inside the image. Here is what I am using to test it:

here is the content of `Dockerfile`:
```
FROM node:18.14.1-alpine
WORKDIR /app
COPY package*.json ./
RUN npm install
RUN ls -d node_modules/nx*/
```

```
docker build . -t testing --progress=plain --no-cache --platform linux/amd64
#1 [internal] load build definition from Dockerfile
#1 sha256:c60ebfa77bb0c255c4bf039800821e79f9ed127e7b5917d9168436072241c91b
#1 transferring dockerfile: 228B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:1248faf17fb6e753daf5dbcdea50cee4242a4403fb0892f8c1d8aeb336857315
#2 transferring context: 35B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/node:18.14.1-alpine
#3 sha256:4f4b31ec22f05d3e4aff615d4fc09a97cebee998fc6cd136a853767b34a35b09
#3 DONE 0.7s

#4 [1/5] FROM docker.io/library/node:18.14.1-alpine@sha256:e0a779479fca9cf43cae2852291bee70e730ae3ad27fea1211060bd344b696b8
#4 sha256:34ea6150a404a947870f9ef41d5ad895932859fccf96472fb391e0f32535504f
#4 DONE 0.0s

#5 [2/5] WORKDIR /app
#5 sha256:253858d1948a7c30557fbf9e1f5ee06be0aaee31c3dc301c810156a8f6f6021b
#5 CACHED

#6 [internal] load build context
#6 sha256:07d9588fa6c5c6227e76d3f7d676a123f1671c7a8ca5fdc533891024d7e8d963
#6 transferring context: 34B done
#6 DONE 0.0s

#7 [3/5] COPY package*.json ./
#7 sha256:de89b53a98249277d5c52bd73871392735486669aed4c74b98a316c8ea279a51
#7 DONE 0.0s

#8 [4/5] RUN npm install
#8 sha256:a4f48f2984d27239ea162955d80e6fa71bf5ca1b6a504b63b876e0a5ba10f509
#8 119.1 npm WARN ERESOLVE overriding peer dependency
#8 343.3 npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
#8 343.5 npm WARN deprecated source-map-resolve@0.6.0: See https://github.com/lydell/source-map-resolve#deprecated
#8 348.3 npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
#8 372.6 npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
#8 397.8 npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
#8 463.5
#8 463.5 added 2360 packages, and audited 2361 packages in 8m
#8 463.5
#8 463.5 284 packages are looking for funding
#8 463.5   run `npm fund` for details
#8 463.5
#8 463.5 7 vulnerabilities (2 moderate, 5 high)
#8 463.5
#8 463.5 To address all issues (including breaking changes), run:
#8 463.5   npm audit fix --force
#8 463.5
#8 463.5 Run `npm audit` for details.
#8 463.5 npm notice
#8 463.5 npm notice New minor version of npm available! 9.3.1 -> 9.6.1
#8 463.5 npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.6.1>
#8 463.5 npm notice Run `npm install -g npm@9.6.1` to update!
#8 463.5 npm notice
#8 DONE 466.8s

#9 [5/5] RUN ls -d node_modules/nx*/
#9 sha256:4d86d55109ef19e0042505c19e59521df99f3fb681607430d972056c65eea1be
#9 0.335 node_modules/nx/
#9 DONE 0.4s

#10 exporting to image
#10 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
#10 exporting layers
#10 exporting layers 29.4s done
#10 writing image sha256:2af49d419f4bcffea5fa8fc8b54b262ce14f20f719c196b30b1872bf33b1ad43
#10 writing image sha256:2af49d419f4bcffea5fa8fc8b54b262ce14f20f719c196b30b1872bf33b1ad43 done
#10 naming to docker.io/library/testing done
#10 DONE 29.4s
```

BTW: I tried to delete `package-lock.json`, but no difference.

--- a/docs/generated/manifests/menus.json
+++ b/docs/generated/manifests/menus.json
@@ -1973,6 +1973,14 @@
             "isExternal": false,
             "children": [],
             "disableCollapsible": false
+          },
+          {
+            "name": "Troubleshooting Nx Install Issues",
+            "path": "/recipes/ci/troubleshoot-nx-install-issues",
+            "id": "troubleshoot-nx-install-issues",
+            "isExternal": false,
+            "children": [],
+            "disableCollapsible": false
           }
         ],
         "disableCollapsible": false
@@ -2037,6 +2045,14 @@
         "name": "Setting up Bitbucket",
         "path": "/recipes/ci/monorepo-ci-bitbucket-pipelines",
         "id": "monorepo-ci-bitbucket-pipelines",
+        "isExternal": false,
+        "children": [],
+        "disableCollapsible": false
+      },
+      {
+        "name": "Troubleshooting Nx Install Issues",
+        "path": "/recipes/ci/troubleshoot-nx-install-issues",
+        "id": "troubleshoot-nx-install-issues",
         "isExternal": false,
         "children": [],
         "disableCollapsible": false

--- a/docs/generated/manifests/recipes.json
+++ b/docs/generated/manifests/recipes.json
@@ -246,6 +246,16 @@
         "isExternal": false,
         "path": "/recipes/ci/monorepo-ci-bitbucket-pipelines",
         "tags": []
+      },
+      {
+        "id": "troubleshoot-nx-install-issues",
+        "name": "Troubleshooting Nx Install Issues",
+        "description": "",
+        "file": "shared/installation/troubleshoot-installation",
+        "itemList": [],
+        "isExternal": false,
+        "path": "/recipes/ci/troubleshoot-nx-install-issues",
+        "tags": []
       }
     ],
     "isExternal": false,
@@ -330,6 +340,16 @@
     "itemList": [],
     "isExternal": false,
     "path": "/recipes/ci/monorepo-ci-bitbucket-pipelines",
+    "tags": []
+  },
+  "/recipes/ci/troubleshoot-nx-install-issues": {
+    "id": "troubleshoot-nx-install-issues",
+    "name": "Troubleshooting Nx Install Issues",
+    "description": "",
+    "file": "shared/installation/troubleshoot-installation",
+    "itemList": [],
+    "isExternal": false,
+    "path": "/recipes/ci/troubleshoot-nx-install-issues",
     "tags": []
   },
   "/recipes/adopting-nx": {

--- a/docs/map.json
+++ b/docs/map.json
@@ -809,6 +809,11 @@
               "name": "Setting up Bitbucket",
               "id": "monorepo-ci-bitbucket-pipelines",
               "file": "shared/monorepo-ci-bitbucket-pipelines"
+            },
+            {
+              "name": "Troubleshooting Nx Install Issues",
+              "id": "troubleshoot-nx-install-issues",
+              "file": "shared/installation/troubleshoot-installation"
             }
           ]
         },

--- a/docs/shared/installation/troubleshoot-installation.md
+++ b/docs/shared/installation/troubleshoot-installation.md
@@ -4,8 +4,6 @@ Here are some common scenarios we came across while trying to run Nx on CIs
 
 ## Native Modules
 
-<!-- NX   Cannot find module '@nrwl/nx-linux-x64-gnu' -->
-
 With more recent versions of Nx, we publish native binaries that should be automatically downloaded and installed when Nx is used.
 
 Some npm users are experiencing errors like the following:

--- a/docs/shared/installation/troubleshoot-installation.md
+++ b/docs/shared/installation/troubleshoot-installation.md
@@ -1,0 +1,46 @@
+# Troubleshoot Nx Installations
+
+Here are some common scenarios we came across while trying to run Nx on CIs
+
+## Native Modules
+
+<!-- NX   Cannot find module '@nrwl/nx-linux-x64-gnu' -->
+
+With more recent versions of Nx, we publish native binaries that should be automatically downloaded and installed when Nx is used.
+
+Some npm users are experiencing errors like the following:
+
+```shell
+NX   Cannot find module '@nrwl/nx-linux-x64-gnu'
+```
+
+This happens because the package-lock.json file was not correctly updated by npm, and missed optional dependencies used by Nx.
+You can read more about this [issue on the npm repository.](https://github.com/npm/cli/issues/4828)
+
+{% callout type="note" title="Updating Nx" %}
+When updating Nx that is already on 15.8, the package-lock.json should continue to be updated properly with all the proper optional dependencies.
+{% /callout %}
+
+### How to fix
+
+Deleting your node_modules and `package-lock.json` and re-running `npm i` should make sure that everything is created properly in the newly generated `package-lock.json`.
+
+### Supported native module platforms
+
+We publish modules for the following platforms:
+
+- macOS 11+ (arm64, x64)
+- Windows (arm64, x64)
+  - We use the `msvc` target, so as long as Microsoft supports your Windows version, it should work on it
+- Linux (arm64, x64)
+  - We use `gnu` ang `musl` targets, which are used by the most popular Linux distributions
+
+If you're running a machine that isn't part of the list above, then Nx will fall back to a non-native implementation where needed.
+
+{% callout type="caution" title="Hash mismatches" %}
+One of the places where we use native modules is for calculating file hashes. The native implementation uses xxHash, while the fallback implementations use sha.
+
+If your CI and local machines are using different implementations (e.g. your CI is using a machine that isn't supported), cache hits will not match between local and CI runs.
+
+You can opt out of using the native hasher by having `NX_NON_NATIVE_HASHER=true` set as an environment variable.
+{% /callout %}

